In [1]:
#pip install statsmodels
#conda install xgboost


In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import scipy.stats as stats
from scipy.stats import ttest_ind
import statsmodels.stats.power as smp
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import xgboost as xgb

In [3]:
data = pd.read_csv('House Prices.csv')

In [4]:
data.head()

date      price  bedrooms  bathrooms  sqft_living  sqft_lot  \
0  2014-05-02 00:00:00   313000.0       3.0       1.50         1340      7912   
1  2014-05-02 00:00:00  2384000.0       5.0       2.50         3650      9050   
2  2014-05-02 00:00:00   342000.0       3.0       2.00         1930     11947   
3  2014-05-02 00:00:00   420000.0       3.0       2.25         2000      8030   
4  2014-05-02 00:00:00   550000.0       4.0       2.50         1940     10500   

   floors  waterfront  view  condition  sqft_above  sqft_basement  yr_built  \
0     1.5           0     0          3        1340              0      1955   
1     2.0           0     4          5        3370            280      1921   
2     1.0           0     0          4        1930              0      1966   
3     1.0           0     0          4        1000           1000      1963   
4     1.0           0     0          4        1140            800      1976   

   yr_renovated                    street       city  statezip country  
0          2005      18810 Densmore Ave N  Shoreline  WA 98133     USA  
1             0           709 W Blaine St    Seattle  WA 98119     USA  
2             0  26206-26214 143rd Ave SE       Kent  WA 98042     USA  
3             0           857 170th Pl NE   Bellevue  WA 98008     USA  
4          1992         9105 170th Ave NE    Redmond  WA 98052     USA

In [5]:
data.isna()

date  price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0     False  False     False      False        False     False   False   
1     False  False     False      False        False     False   False   
2     False  False     False      False        False     False   False   
3     False  False     False      False        False     False   False   
4     False  False     False      False        False     False   False   
...     ...    ...       ...        ...          ...       ...     ...   
4595  False  False     False      False        False     False   False   
4596  False  False     False      False        False     False   False   
4597  False  False     False      False        False     False   False   
4598  False  False     False      False        False     False   False   
4599  False  False     False      False        False     False   False   

      waterfront   view  condition  sqft_above  sqft_basement  yr_built  \
0          False  False      False       False          False     False   
1          False  False      False       False          False     False   
2          False  False      False       False          False     False   
3          False  False      False       False          False     False   
4          False  False      False       False          False     False   
...          ...    ...        ...         ...            ...       ...   
4595       False  False      False       False          False     False   
4596       False  False      False       False          False     False   
4597       False  False      False       False          False     False   
4598       False  False      False       False          False     False   
4599       False  False      False       False          False     False   

      yr_renovated  street   city  statezip  country  
0            False   False  False     False    False  
1            False   False  False     False    False  
2            False   False  False     False    False  
3            False   False  False     False    False  
4            False   False  False     False    False  
...            ...     ...    ...       ...      ...  
4595         False   False  False     False    False  
4596         False   False  False     False    False  
4597         False   False  False     False    False  
4598         False   False  False     False    False  
4599         False   False  False     False    False  

[4600 rows x 18 columns]

In [6]:
data['price'].mean()

551962.9884732141

# Hypothesis Testing

Performing Hypothisis testing for the Alpha at 5% for the house prices that house mean house price value of $551963 if the P value is for the alpha of 0.05 is less then we will refect the null hypothesis

In [7]:
alpha = 0.5
null_hypothesis = data['price'].mean()


In [8]:
# Here we are performing T-Test for the hypothesis testing
t_stat, p_value  = stats.ttest_1samp(data['price'],null_hypothesis)

In [9]:
if p_value<alpha :
    print("Reject the Null Hypothesis")
else:
    print("Fail to reject null hypothesis")

Fail to reject null hypothesis


Here we can see that we fail to reject the Null hypothesis , that means our house price data is consistent with an assumption that the population mean is equal to $551962.9884732141

# Alpha/ Beta Testing

In [10]:
grp_a = data['price'].iloc[:2300]
grp_b = data['price'].iloc[2300:]

In [11]:
grp_a

0        313000.0
1       2384000.0
2        342000.0
3        420000.0
4        550000.0
          ...    
2295     200000.0
2296     295000.0
2297     580379.0
2298     875000.0
2299     515000.0
Name: price, Length: 2300, dtype: float64

In [12]:
grp_b

2300    599000.000000
2301    285000.000000
2302    290000.000000
2303    355000.000000
2304    381000.000000
            ...      
4595    308166.666667
4596    534333.333333
4597    416904.166667
4598    203400.000000
4599    220600.000000
Name: price, Length: 2300, dtype: float64

In [13]:
t_stat1,p_value1 = ttest_ind(grp_a,grp_b)
print("T-statistics",t_stat1)
print("P-Value",p_value1)


T-statistics -0.6048022124721948
P-Value 0.5453402677120369


In [14]:
#statistical Power in order to calculate the difference between two groups
# n = len(data)
# effect_size = np.mean(grp_a) - np.mean(grp_b)
# alpha = 0.05
# ratio = 1


In [15]:
# # power of the test
# power = smp.TTestIndPower().power(effect_size=effect_size,nobs1=n*ratio,alpha = alpha,ratio = ratio,alternative="two-sided")
# print("Power of the test",power)

# Machine Learning Models like Regression, Classification

In [16]:
data.head()

date      price  bedrooms  bathrooms  sqft_living  sqft_lot  \
0  2014-05-02 00:00:00   313000.0       3.0       1.50         1340      7912   
1  2014-05-02 00:00:00  2384000.0       5.0       2.50         3650      9050   
2  2014-05-02 00:00:00   342000.0       3.0       2.00         1930     11947   
3  2014-05-02 00:00:00   420000.0       3.0       2.25         2000      8030   
4  2014-05-02 00:00:00   550000.0       4.0       2.50         1940     10500   

   floors  waterfront  view  condition  sqft_above  sqft_basement  yr_built  \
0     1.5           0     0          3        1340              0      1955   
1     2.0           0     4          5        3370            280      1921   
2     1.0           0     0          4        1930              0      1966   
3     1.0           0     0          4        1000           1000      1963   
4     1.0           0     0          4        1140            800      1976   

   yr_renovated                    street       city  statezip country  
0          2005      18810 Densmore Ave N  Shoreline  WA 98133     USA  
1             0           709 W Blaine St    Seattle  WA 98119     USA  
2             0  26206-26214 143rd Ave SE       Kent  WA 98042     USA  
3             0           857 170th Pl NE   Bellevue  WA 98008     USA  
4          1992         9105 170th Ave NE    Redmond  WA 98052     USA

In [17]:
X = data.drop(['date','price','street','city','statezip','country'],axis = 1)
y = data['price']

In [18]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=42)


In [19]:
lr = LinearRegression()
lr.fit(X_train,y_train)

LinearRegression()

In [20]:
prediction = lr.predict(X_test)
prediction

array([ 304863.38578053,  326517.06991506, 1071762.53630093,
        546241.40097351,  379799.98321077,  605792.93622748,
        481263.5712065 ,  418905.11058129,  509341.50791915,
        530852.97588251,  682958.97641215,  414525.3613005 ,
        836974.40259906,  416500.85730824,  367263.43388217,
        702908.51672183,  677878.69067825,  515539.32998912,
       1019380.63891402,  861119.79490726, 1364866.21814694,
        637439.70016772,  634739.56666336,  470424.21250257,
        162037.15383025,  232831.37648233,  669951.05071601,
        884080.66695004,  263027.22577195,  974954.64229475,
       1883342.16391336,  481510.29664849, 1275554.6525372 ,
        430994.93178858,  180366.40705409,  339240.17357412,
        792335.74094022, 1011053.14613369,  238036.2092767 ,
        543438.66040774,  423414.50597284,  243429.92107964,
        390151.3007404 ,  361322.67375674,  320343.34636007,
        317865.29287637,  469559.60646618,  573249.20158967,
        828966.75691609,

In [21]:
mse = np.mean((prediction - y_test)**2)
print("Mean Squared Error",mse)

Mean Squared Error 986921767056.1467


In [22]:
accurancy = lr.score(X_test,y_test)
accurancy

0.03228385663275568

In [26]:
# Gradient boosting
X1 = data.drop(['date','price','street','city','statezip','country'],axis =1)
y1 = data['price']


In [27]:
X1_train, X1_test, y1_train, y1_test = train_test_split(X1, y1, test_size=0.2, random_state=42)


In [28]:
xgb_model = xgb.XGBRegressor()
xgb_model.fit(X1_train, y1_train)

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=None, gpu_id=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=None, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=None, max_leaves=None,
             min_child_weight=None, missing=nan, monotone_constraints=None,
             n_estimators=100, n_jobs=None, num_parallel_tree=None,
             predictor=None, random_state=None, ...)

In [29]:
y1_pred = xgb_model.predict(X1_test)

In [30]:
mse1 = mean_squared_error(y1_test,y1_pred)
mse

986921767056.1467

In [33]:
xgb_model.score(y1_pred,X1_test)

XGBoostError: [19:04:47] C:\Users\dev-admin\croot2\xgboost-split_1675461376218\work\src\predictor\cpu_predictor.cc:377: Check failed: m->NumColumns() == model.learner_model_param->num_feature (1 vs. 12) : Number of columns in data must equal to trained model.